File: hamming-error-correction/src/test/java/com/pivovarit/hamming/vavr/HammingPropertyTest.java
package com.pivovarit.hamming.vavr;

import io.vavr.collection.Stream;
import org.junit.jupiter.api.Test;

import java.util.Random;

import static org.assertj.core.api.Assertions.assertThat;

class HammingPropertyTest {

    private static final Random rand = new Random();
    private final HammingEncoder encoder = HammingEncoder.vavrEncoder();
    private final HammingDecoder decoder = HammingDecoder.vavrDecoder();

    @Test
    void shouldEncodeAndDecode() {
        for (int i = 0; i < 10000; i++) {
            BinaryString msg = randomMessage();
            assertThat(msg)
              .isEqualTo(decoder.decode(encoder.encode(msg)));
        }
    }

    @Test
    void shouldEncodeAndDecodeWithSingleBitErrors() {
        for (int i = 0; i < 10000; i++) {
            BinaryString msg = randomMessage();
            EncodedString encoded = encoder.encode(msg);
            assertThat(msg)
              .isEqualTo(decoder.decode(withBitFlippedAt(rand.nextInt(encoded.getValue().length()), encoded)));
        }
    }

    private BinaryString randomMessage() {
        String msg = Stream.continually(() -> rand.nextInt(2))
          .map(i -> Integer.toString(i))
          .take(rand.nextInt(10) + 1)
          .reduce(String::concat);

        return BinaryString.of(msg);
    }

    private EncodedString withBitFlippedAt(int ind, EncodedString source) {
        char it = source.getValue().charAt(ind);
        StringBuilder builder = new StringBuilder(source.getValue());
        builder.setCharAt(ind, it == '0' ? '1' : '0');
        return EncodedString.of(builder.toString());
    }
}

File: hamming-error-correction/src/test/kotlin/com/pivovarit/hamming/domain/HammingPropertyTest.kt
package com.pivovarit.hamming.domain

import com.pivovarit.hamming.domain.decode.HammingDecoder
import com.pivovarit.hamming.domain.encode.HammingEncoder
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.Random

class HammingPropertyTest {
    private val rand = Random()

    private val encoder = HammingEncoder.sequentialStateless()
    private val decoder = HammingDecoder.sequentialStateless()

    @Test
    fun shouldEncodeAndDecode() = repeat(10000) {
        randomMessage().let {
            assertThat(it)
              .isEqualTo(decoder.decode(encoder.encode(it)))
        }
    }

    @Test
    fun shouldEncodeAndDecodeWithSingleBitErrors() = repeat(10000) {
        randomMessage().let {
            assertThat(it).isEqualTo(decoder.decode(encoder.encode(it)
              .withBitFlippedAt(rand.nextInt(it.length))))
        }
    }

    private fun randomMessage(): BinaryString =
      generateSequence { rand.nextInt(2).toString() }
        .take(rand.nextInt(1000).inc())
        .reduce { acc, s -> acc + s }
        .let(::BinaryString)


    private fun EncodedString.withBitFlippedAt(ind: Int): EncodedString = this[ind].toString().toInt()
      .let { this.value.replaceRange(ind, ind + 1, ((it + 1) % 2).toString()) }
      .let(::EncodedString)
}
